### ISSUE TITLE
How to reuse old drawable in a GifDrawable subclass

### ISSUE THREAD

**Issue by storix** [koral--/android-gif-drawable#398]
It looks like there is no way to reuse the old `GifDrawable` in a `GifDrawable` subclass. Maybe I missed something? I looked through the code and there is no public `GifDrawable` constructor which accepts `oldDrawable`. I cannot subclass `GifDrawableBuilder` for this purpose either.

**Comment by koral--** [comment]
Currently there is no way to do that it is by design. Constructor accepting old `GifDrawable` also uses `GifInfoHandle` which was not designed to be exposed. 
I'll consider adding protected constructor so I'm keeping this issue open.

Of course feel free to send pull request with that.

**Comment by storix** [comment]
I forked the project and made the following changes:

1. The `GifDrawable` constructor which accepts `oldDrawable` has been made [protected](https://github.com/storix/android-gif-drawable/commit/5dd3c1ec96c72b4b72a32823e3b5caddb4e68e29).

2. The `GifInfoHandle` constructors visibility has been changed to [public](https://github.com/storix/android-gif-drawable/commit/063262298bf693daf1dbe22a0dc3c28078323ad5). I understand that you may find this as inappropriate but without public `GifInfoHandle` we would need to duplicate all the `GifDrawable` constructors to make them accept `oldDrawable`. 

3. The `GifDrawableBuilder` has been [refactored](https://github.com/storix/android-gif-drawable/commit/c20e9e5fbc4b7b1afd0815a875bff7c0c03dbdf9) to allow subclassing.  I followed the design pattern proposed in [this](https://community.oracle.com/blogs/emcmanus/2010/10/24/using-builder-pattern-subclasses) article. 

     The advantages of the pattern:

    - There are no breaking changes to the existing code base
    - Type safety (there are no `unchecked cast` warnings)

    Cons:
    - A little bit of boilerplate since subclasses will have to define the following method to ensure type safety:
```Java
        @Override
        protected Builder self() {
            return this;
        }
```

4. To simplify the builder subclassing I also added [GifDrawableFactory](https://github.com/storix/android-gif-drawable/commit/a498a8d5af13caabee7d983e028fa44ab1f4b32f) and a new [public InputSource build method ](https://github.com/storix/android-gif-drawable/commit/97cbdac1c24fd348f46d8f994c15ef2351c456bb) which accepts the concrete instances of the factory.
___________

At the usage site it looks like this:

- Java

```Java
public class CustomGifDrawable extends GifDrawable {
    public CustomGifDrawable(final GifInfoHandle handle, final GifDrawable oldDrawable) {
        super(handle, oldDrawable, null, true);
    }

    protected abstract static class CustomGifDrawableInit<T extends CustomGifDrawableInit<T>> extends GifDrawableInit<T> {
        //Add new builder options here

        @Override
        public CustomGifDrawable build() throws IOException {
            final InputSource inputSource = getInputSource();
            checkNotNull(inputSource, "Input source must not be null");

            return inputSource.build(getOldDrawable(), getExecutor(), isRenderingTriggeredOnDraw(), getOptions(), new GifDrawableFactory<CustomGifDrawable>() {
                @Override
                public CustomGifDrawable create(GifInfoHandle gifInfoHandle, GifDrawable oldDrawable, ScheduledThreadPoolExecutor executor, boolean isRenderingTriggeredOnDraw) {
                    return new CustomGifDrawable(gifInfoHandle, oldDrawable);
                }
            });
        }
    }

    public static class CustomGifDrawableBuilder extends CustomGifDrawableInit<CustomGifDrawableBuilder> {
        // The boilerplate for type safety
        @Override
        protected CustomGifDrawableBuilder self() {
            return this;
        }
    }
}
```

- Kotlin

```Kotlin
class CustomGifDrawable(handle: GifInfoHandle, oldDrawable: GifDrawable?) : GifDrawable(handle, oldDrawable, null, true) {

    abstract class CustomGifDrawableInit<T : CustomGifDrawableInit<T>> : GifDrawableInit<T>() {
        override fun build(): CustomGifDrawable {
            checkNotNull(inputSource, { "Input source must not be null" })

            return inputSource.build(oldDrawable, executor, isRenderingTriggeredOnDraw, options, {
                handle, oldDrawable, _, _ ->
                CustomGifDrawable(handle, oldDrawable)
            })
        }
    }

    class CustomGifDrawableBuilder : CustomGifDrawableInit<CustomGifDrawableBuilder>() {
        override fun self(): CustomGifDrawableBuilder = this
    }
}
```

Please review the changes when you will have free time and if you find it acceptable I'll make PR.

**Comment by koral--** [comment]
Looks good to me, go ahead with PR.

**Comment by koral--** [comment]
Released in version `1.2.7`
